### ISSUE TITLE
"perm" expressions should be forbidden from occurring in function and predicate bodies

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#106]
> Created by **@alexander_summers** on 2015-11-25 15:36, last updated on 2017-08-25 14:42

We agreed this in a Viper meeting, but it is not yet implemented.

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2016-06-23 06:39

They already are rejected in function bodies, but not yet in predicate bodies.

**Comment by fpoli** [comment]
> **@dohrau** commented on 2016-09-14 12:53

Also, they should be rejected in invariants, pre- and postconditions.

**Comment by fpoli** [comment]
> **@vakaras** commented on 2016-09-14 13:02

@dohrau: Are you suggesting to forbid this:

```
#!silver
method send(_cthread: Ref, _caller_measures: Seq[Measure$], t1: Ref, socket: Ref, data: Ref) returns (_res: Ref)
  requires [acc(MustInvokeBounded(t1), write), ((perm(MustInvokeBounded(t1)) > none) && Measure$check(_caller_measures, t1, 1) ? acc(MustInvokeBounded(t1), write) : (perm(MustInvokeUnbounded(t1)) > none ? acc(MustInvokeUnbounded(t1), write) : acc(MustInvokeCredit(t1), write)))] && (acc(send_io(t1, socket, data), write) && ((data != null) && ((socket != null) && [acc(MustTerminate(_cthread), write), true])))

```

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2016-09-14 13:09

We could consider allowing perm expressions in these positions *if* they occur inside inhale-exhale expressions. But otherwise, I am inclined to forbid them, since they are typically unsound (they get exhaled in a state with different permissions available than that in which they get inhaled). This is the same reason we don't want to allow them in e.g. predicates, I think.

**Comment by fpoli** [comment]
> **@vakaras** commented on 2016-09-14 13:41

I see your point. I suggest that before implementing any changes, we create a list of places in which we want to allow  `perm` (and `forperm`?). For example, Nagini and Chalice2Viper uses `perm` and `forperm` in (writing from the top of my head, list could be non-exhaustive):

1. Preconditions, postconditions and loop invariants, but only in exhale. (For switching between obligation exhale phases and for leak checks.)
2. Expressions assigned to variables: `currentTerminateAmount := perm(MustTerminate(_cthread))`. (For storing current permission amount.)
3. In inhale statements: `inhale acc(MustTerminate(_cthread), currentTerminateAmount - perm(MustTerminate(_cthread)))`. (To restore previous permission amount.)
4. In exhale statements: `exhale acc(MayReceiveGuard(_cthread), perm(MayReceiveGuard(_cthread)))`. (To zero permission amount.)

Note: this is what is used by current implementation of Nagini/Chalice2Viper. For my current design of deadlock avoidance (which is not implemented yet), I also need `forperm` in postcondition inhale.

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2017-06-16 07:25

See also [Silicon issue #291](https://bitbucket.org/viperproject/silicon/issues/291)

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2017-08-25 14:42

Resolved by @mousam05 with pull request #73.

What's currently implemented corresponds to @vakaras' last comment, plus `perm` and `forperm` are completely forbidden in function and predicate bodies (even if nested under inhale-exhale expressions). 

See `all/issues/silver/0106.sil` and `0106-1.sil` for corresponding regression tests.
